lib-classifier: refactor single image viewer with VisXZoom

mcbouslog opened the pull request:
## Package
- lib-classifier

## Linked Issue and/or Talk Post
- closes #5109 

## Describe your changes
- [x] create new single image subject viewer
- [x] create storybook for new single image subject viewer
- [x] rotate
- [x] invert
- [x] add VisXZoom for pan and zoom (**includes mouse wheel zoom**)
- [x] drag
- [x] drawing / InteractionLayer 
- [x] placeholder image
- [x] check new single image subject viewer in test project
- [x] check new single image subject viewer with other viewers that use single image viewer
- [x] fix broken tests
- [x] add new tests
- [x] update related READMEs (i.e. [Subject Viewer](https://github.com/zooniverse/front-end-monorepo/tree/master/packages/lib-classifier/src/components/Classifier/components/SubjectViewer#pan-and-zoom), likely others)
- ~review mobile/pinch handling~ new functionality will be investigated and potentially a separate PR

## Notes
- see existing VisXZoom notes - https://github.com/zooniverse/front-end-monorepo/tree/master/packages/lib-classifier/src/components/Classifier/components/SubjectViewer#pan-and-zoom
- ZoomingScatterPlot uses VisXZoom: https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/ScatterPlotViewer/components/ZoomingScatterPlot/ZoomingScatterPlot.js#L206-L224

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX?
  - test project with single and multi image subjects, drawing task: 
    - staging/lib-classifier: https://local.zooniverse.org:8080/?project=1693&workflow=3838
    - staging/app-project: https://local.zooniverse.org:3000/projects/markb-panoptes/test-project-1-mb-fem-lab
- What user actions should my reviewer step through to review this PR?
  1. load subject with image(s)
  2. pan, zoom, rotate, invert, and reset image
  3. draw marks on the image, repeat step 2 with drawn marks
- Which storybook stories should be reviewed?
  - mainly http://localhost:6006/?path=/story/subject-viewers-singleimageviewer--default, but also DataImageViewer, FlipbookViewer, ImageAndTextViewer, MultiFrameViewer, and SeparateFrames stories
  - SeparateFrames stories onKeyDown (i.e. `+` to zoom in) doesn't work, but it does in the linked test project, not sure why
- Are there plans for follow up PR’s to further fix this bug or develop this feature?

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

mcbouslog commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageCanvas.js:
```diff
@@ -0,0 +1,141 @@
+
+import SVGImage from '../SVGComponents/SVGImage'
+
+function calculateAdjustedViewBox({ naturalWidth, naturalHeight, transformMatrix }) {
```
I think this calculation may need additional work, specifically to handle zoomed out panning. When zooming out, then panning the image jumps around erratically. I'm looking into the issue, but suggestions welcome.

mcbouslog replied:
Fixed in [4760ea4](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/4760ea4879469b0f220982d15f810b595c618dc9)

mcbouslog commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.js:
```diff
@@ -1,123 +1,126 @@
+        {title?.id && title?.text && (
+          <title id={title.id}>{title.text}</title>
+        )}
+          <VisXZoom
+            height={naturalHeight}
+            panning={panning}
+            setOnPan={setOnPan}
+            setOnZoom={setOnZoom}
+            width={naturalWidth}
+            zoomConfiguration={DEFAULT_ZOOM_CONFIG}
+            zoomingComponent={SingleImageCanvas}
+            zooming={zooming}
+            {...singleImageCanvasProps}
+          />
```
See https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/ScatterPlotViewer/components/ZoomingScatterPlot/ZoomingScatterPlot.js#L206-L224 for existing implementation of VisXZoom. `constrain` could be added to the SingleImageViewer potentially.

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/README.md:
```diff
@@ -36,12 +36,11 @@ A subject viewer is a component designed to render the media of a subject and an
 
-- [`VisXZoom`](components/SVGComponents/VisXZoom) - An extension of the VisX library's pan and zoom functional component. Uses a transformation matrix, boolean control for zoom and pan individual to turn the functions on and off as needed, has configurability for zoom direction, minimum zoom, maximum zoom, zoom in value, and zoom out value, and supports zoom to point. Currently implemented with t ...
-- [`SVGPanZoom`](components/SVGComponents/SVGPanZoom) - A port of the PFE pan and zoom functionality being used with the `SingleImageViewer` and subsequently the `MultiFrameViewer` and anywhere else the `SingleImageViewer` may be used. It does scale transformations on the SVG view box. 
+- [`VisXZoom`](components/SVGComponents/VisXZoom) - An extension of the VisX library's pan and zoom functional component. Uses a transformation matrix, boolean control for zoom and pan individual to turn the functions on and off as needed, has configurability for zoom direction, minimum zoom, maximum zoom, zoom in value, and zoom out value, and supports zoom to point. Currently implemented with t ...
```
Just out of curiosity is the 'todo' at the end in a Github Issue or referenced somewhere else in the code?

mcbouslog replied:
I think #2110 is the related issue. I'm not sure the NOTE belongs in the README though. I could remove the NOTE or update it to include the specific issue?

mcbouslog replied:
Removed TODO in [7dc2905](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/7dc29051c505a72ed7b00f38a16cf87ab3913d61)

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/FlipbookViewer.js:
```diff
@@ -59,6 +37,18 @@ const FlipbookViewer = ({
+  const imageLocationUrl = subject?.locations[currentFrame]?.url
+
+  const { img, error, loading, subjectImage } = useSubjectImage({
+    src: imageLocationUrl,
```
Changing this from `defaultFrameLocation` to `imageLocationUrl` means `useSubjectImage()` hook is called on every frame because `imageLocationUrl` is determined by `currentFrame` rather than `defaultFrame`. This is undesired behavior because it directly calls `onSubjectReady()` every time a volunteer flips to a new frame. `onSubjectReady()` resets the rotation between frames and pushes new metadata to the classification.

Regarding `rotation`, we want volunteers to be able to rotate the subject and 'play' the flipbook with a subject rotated. `rotation` should not be reset in between every frame.

I noticed the metadata bug because earlier this year it was introduced by https://github.com/zooniverse/front-end-monorepo/pull/5754 and fixed by https://github.com/zooniverse/front-end-monorepo/pull/5952. On this branch, when I used my flipbook test project (listed in the FEM [active projects](https://docs.google.com/spreadsheets/d/1X9345vQ_xNX6BJf9_xwVCQCjxjbWWErGmV7Xe8WpHZo/edit?gid=223454926#gid=223454926) for reference), this is what a submitted classification looks like when I flipped through 12 frames. Huge metadata causes classification exports to hang.
<img width="458" alt="Screenshot 2024-11-18 at 3 32 35 PM" src="https://github.com/user-attachments/assets/159e1f1d-0511-451d-9176-f30d41c5941d">
___
Questions - Is the change from `defaultFrameLocation` to `imageLocationUrl` intentional and required for the new SingleImageViewer with VisXZoom? How can each viewer ensure `onSubjectReady()` is called only once per subject, and not per frame? I didn't look into ImageAndTextViewer, MultiFrameViewer, and SeparateFramesViewer in detail yet, but I want to include those in this discussion too.

mcbouslog replied:
> Questions - Is the change from defaultFrameLocation to imageLocationUrl intentional and required for the new SingleImageViewer with VisXZoom?

I don't think it's required. I'll try refactoring so `onSubjectReady` (`onReady` passed from `SubjectViewer`) is only called once per subject and report back.

mcbouslog replied:
I believe `classification.metadata.subject_dimensions` are now consistent per [c58b452](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/c58b452d867b7bad29688b9a515573121512ece7) and [7eb30ef](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/7eb30eff5c4e2fa32398947b67e65419701417ec) and [0a2d54f](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/0a2d54fae488d27c0746d4fb1000af9ceaf8e5ec).

goplayoutside3 reviewed (commented):
I started reading through each viewer and testing their UX. Because this will be an in depth review and lots of moving parts, I'm pausing with a couple of questions and will resume once they're resolved.

eatyourgreens commented:
When Sarah converted the single image viewer to `VisXZoom` (back in 2020, I think), it broke all the drawing tools. This PR breaks all the drawing tools too, in (I think) the same way. Should I add videos/issue descriptions, or is this still a work in progress?

eatyourgreens commented:
There's an undocumented feature, used by Seabird Watch and Penguin Watch, which you should add to your manual drawing tool tests.

When you're making a large number of marks, you don't want to be constantly switching between zoom mode and drawing mode. For this reason, you can control pan-and-zoom with the keyboard, from a focused drawing mark.

To see this in action, try placing marks, then zooming and panning with the keyboard at either of the following links:
- https://www.zooniverse.org/projects/penguintom79/penguin-watch/classify
- https://frontend.preview.zooniverse.org/projects/penguintom79/penguin-watch/classify

You should be able to mark a penguin, then zoom and pan from the keybard while continuing to mark penguins on the zoomed picture (on crowded pictures, you really need to zoom in, so this feature isn't optional.)

This PR breaks that keyboard accessibility for drawn marks.

eatyourgreens commented:
One last comment: on this branch, drawn marks scale up with the image when you zoom in. Drawn marks should ignore the scale transformation and stay the same size at all zoom levels (see #6066 for one attempt to fix this for stroke widths. You can also style SVG shapes so that they are indepenedent of scale.)

mcbouslog commented:
Thank you @eatyourgreens , the comments and videos are very helpful! To summarize I think you've identified 3 issues:
1. marks with drag and move handles aren't scaled and working as expected
2. keyboard pan and zoom should be enabled in annotation mode
3. marks aren't properly scaled zoomed in or out

I'll report back on progress.

eatyourgreens commented:
>Thank you @eatyourgreens , the comments and videos are very helpful! To summarize I think you've identified 3 issues:
>
>- marks with drag and move handles aren't scaled and working as expected
 
You're welcome. For that first one, I can also reproduce it with the point tool, without drag handles. Drop a point on the image somewhere, then drag and move it **while zoomed in**. The pointer movement, within the viewport, isn't properly mapped to the image. It looks like the pointer coordinates are still scaled to the viewport dimensions, not the new, larger image dimensions.

mcbouslog commented:
I can recreate the noted 2 and 3 issues, but I'm having trouble recreating 1 (marks with drag and move handles aren't scaled and working as expected)? I'm using Chrome v131, but I don't think it'd be browser specific.

https://github.com/user-attachments/assets/0481249c-7946-4ba5-9948-de91358fdcad

eatyourgreens commented:
Sorry, I forgot to say that my video was Firefox 132. SVG implementations do vary between browsers, unfortunately.

eatyourgreens commented:
I tried a few different browsers. Scaling and dragging seems to be ok in Chrome or Safari, but broken in Firefox.

eatyourgreens commented:
This might be relevant. Dragging interactions get the current transformation matrix from a `canvas` element stored in `SVGContext` here:
https://github.com/zooniverse/front-end-monorepo/blob/1862adebef4ddfde8d071669c99038bf3343383c/packages/lib-classifier/src/plugins/drawingTools/components/draggable/draggable.js#L51

Creating a mark here gets the CTM from the `DrawingCanvas` element in the `InteractionLayer` component.
https://github.com/zooniverse/front-end-monorepo/blob/1862adebef4ddfde8d071669c99038bf3343383c/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js#L88

I don't think those are the same element, so that could account for differences in behaviour between creating a mark and dragging a mark. They probably should be using the same element and transformation matrix. Id on't think I've noticed this before. At least, I can't find an issue for it, and I'm sure I would have opened an issue for something like this.

eatyourgreens commented:
Update: dragging a zoomed-in mark is broken in Firefox on the production release. I'll open an issue. It's not a bug that's been introduced on this branch.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageCanvas.js:
```diff
@@ -0,0 +1,141 @@
+          transform={rotationTransform}
+        >
+          <svg
+            ref={canvasLayer}
```
This ref is causing the broken dragging behaviour in Firefox. I've removed it in #6491.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/components/PlaceholderSVG.js:
```diff
@@ -0,0 +1,35 @@
+      maxWidth={maxWidth}
+      tabIndex={0}
+      viewBox={viewBox}
+      xmlns='http://www.w3.org/2000/svg'
```
`xmlns` is ignored by HTML parsers, so you can remove it. See the note at the top of the [MDN SVG element page](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/svg).

mcbouslog replied:
Removed in [e6d5b39](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/e6d5b39b438abcbecb61ec0e8a612928a2ed6afb)

mcbouslog commented:
A brief update on a few issues previously noted:
0. `classification.metadata.subject_dimensions` broken, reversion of previous bug - I think fixed per [c58b452](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/c58b452d867b7bad29688b9a515573121512ece7) and [7eb30ef](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/7eb30eff5c4e2fa32398947b67e65419701417ec) and [0a2d54f](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/0a2d54fae488d27c0746d4fb1000af9ceaf8e5ec)
1. marks with drag and move handles aren't scaled and working as expected - fixed with #6491
2. keyboard pan and zoom should be enabled in annotation mode - fixed with [0193a44](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/0193a44a99bcd1ec1b47ea243e56bdd4f5bf6517)
3. marks aren't properly scaled zoomed in or out - working on
4. ...others to come from additional testing... - working on

mcbouslog commented:
I think this is ready for another review 😵‍💫 !

I've been using https://local.zooniverse.org:3000/projects/markb-panoptes/test-project-1-mb-fem-lab?env=staging&demo=true to test various workflows, image sizes, and drawing tools (see workflow descriptions).

goplayoutside3 commented:
I did a first-pass review using Brave Browser on a Macbook Pro and want to jot down three things for now. I plan to test on another macbook and my phone later on!
1. If the speed of the mouse-wheel zoom is adjustable, I think it should be a little slower. Especially because there's no limit on zooming out, the reaction time to the input from my track pad is jarring. I'll try to find a mouse with an actual wheel to test later too.
2. Looking at metadata per classification, a flipbook subject ends up with more objects than expected in the `metadata.subject_dimensions` specifically when I switch back and forth between separate frames and flipbook.
    - When a subject is loaded in the separate frames view, for instance https://local.zooniverse.org:3000/projects/markb-panoptes/test-project-1-mb-fem-lab/classify/workflow/3429?env=staging&demo=true, there are 2 records of subject dimensions.
    - When a subject is loaded in the flipbook, and a classification submitted without switching to separate frames, there is 1 record of dimensions.
    - By switching back and forth between separate frames and flipbook 6 times, there are 11 records of subject dimensions in the metadata upon classification.
    - Based on this behavior, I think each individual SeparateFrame is somehow signaling `onSubjectReady`. Perhaps there should be a check for a set of dimensions existence in the metadata and don’t add more if it’s already there?
3. We chatted about this in person, but I noticed that when I draw a point on a subject (regardless of viewer) and then rotate the subject 90deg, then zoom out, the size of the point increases. I'm not sure if this is related to `scale` or the stroke width of svg, but I don't believe it affects classification data, just visual differences.

Will continue testing on more devices by end of the week!

eatyourgreens commented:
#6697 should remove the image scaling calculation from this PR. It removes the `scale` prop (and subject image refs) throughout the subject viewers.

eatyourgreens commented:
>Looking at metadata per classification, a flipbook subject ends up with more objects than expected in the metadata.subject_dimensions specifically when I switch back and forth between separate frames and flipbook. 

If it's any help, I'm seeing this bug in #6697 too. It appears to be a bug in the current `production-release`. I can also reproduce it by going to Daily Minor Planet in production, switching to separate frames mode and loading a subject, then switching back and forth between separate frames and flipbook.

eatyourgreens commented:
>Based on this behavior, I think each individual SeparateFrame is somehow signaling onSubjectReady. Perhaps there should be a check for a set of dimensions existence in the metadata and don’t add more if it’s already there?

This works when every frame has the same dimensions, but Squirrel Mapper (which has odd subjects) has subjects where each frame has different dimensions, including subjects where the frames are a mixture of portrait and landscape orientations.

There's also a potential bug when `panoptes-uploads.zooniverse.org` returns errors and subject frames don't load. In that case, it's still possible to submit a Daily Minor Planet classification, even if only one frame loaded. I think the `onSubjectReady` code only requires loading status to be success for one frame, not the whole subject.

mcbouslog commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.js:
```diff
@@ -1,112 +1,129 @@
+  direction: 'both',
+  maxZoom: 10,
+  minZoom: 0.1,
+  onWheelThrottleWait: 100,
```
This throttles the mouse wheel zooming. `100` feels better to me, but open to suggestion!

mcbouslog commented:
I've tested across OS and browsers using BrowserStack and I think this is looking good.

@goplayoutside3 - regarding recent https://github.com/zooniverse/front-end-monorepo/pull/6390#issuecomment-2657744382:
1. mouse wheel zoom speed: decreased frequency of zoom with lodash `throttle` in [89fcf97](https://github.com/zooniverse/front-end-monorepo/pull/6390/commits/89fcf9704f2ce92ef7ee0a2f2ae385f7cce665bf)
2. classification `metadata.subject_dimensions` duplicate entries: fixed in #6581 
3. rotation and drawn mark scaling issue: fixed in #6697

goplayoutside3 reviewed (commented):
1. Things I made sure to manually test based on previous comments. Some of these have been fixed in other PRs, so I just did a quick double check:
    - [x]  Ability to use keyboard pan + zoom while in annotation mode (like Penguin Watch)
    - [x]  Drawn marks map to the image correctly at any zoom level
    - [x]  Expected scale of drawn marks regardless of zoom, rotate, and pan
    - [x]  Expected scale of drag markers (was a bug for Firefox only?)
    - [x]  Metadata’s subject dimensions are recorded once per frame
    - [x]  Mouse-wheel zoom speed is not jarring, it’s comfortable via a touch pad, magic mouse, and with a manual mouse wheel
    - [x]  Storybook functionality of viewers that use SingleImageViewer is as expected.
        - Noted the quirk where an individual SeparateFrame story does not listen to keyboard pan/zoom. `useKeyZoom` looks to have the same setup in that viewer as FlipbookViewer, so I’m not sure why this behavior is specific to Storybook. Keyboard zoom and pan work as expected on a real project run on this branch such as Daily Minor Planet in separate frames view. Maybe related to my note below about unexpected SeparateFramesViewer behavior?

2. Definite Bug

    - In SeparateFramesViewer, I expect clicking an image toolbar always affects the image it belongs to. However, when Daily Minor Planet is run on this branch, the image that is focuses is affected by other image toolbar buttons. For instance, click on the first image, pan or zoom via the toolbar or keyboard, and then click the ‘reset’ button on the second image’s toolbar and the first image resets.
    - I confirmed this bug does not exist on production. Happy to chat in office this week if helpful!

3. Possible bug?

    - In the DataImageViewer, there’s a Story for Pan Zoom where on production the image toolbar buttons can be enabled for the image. However, on this branch the toolbar buttons do not zoom or reset the image in the viewer. The rotate button does work though.
    - I’m not sure which project this viewer was built for, do you know? The viewer itself has several console errors in Storybook which make me think the code hasn’t been touched in a long time. If the bug only exists in Storybook, okay to ignore in this PR, otherwise I’d like to review a live FEM project.

4. Questionable touchscreen behavior?

    - Noting that you mentioned reviewing mobile and pinch handling in the PR description is intended for a separate PR, I only did a quick manual test of looking at [https://local.zooniverse.org:3000/projects/markb-panoptes/test-project-1-mb-fem-lab/classify/workflow/3847](https://local.zooniverse.org:3000/projects/markb-panoptes/test-project-1-mb-fem-lab/classify/workflow/3847?demo=true) on my iPhone. I can draw marks as expected, but when move/pan mode is enabled, dragging an image via touch is clunky and does not respond to the movement of my touch very well. Looking at any other image viewer on production on my phone, dragging is a smoother experience. Is this something you’d like to continue discussing in this PR?

5. Other Notes

    - Panning is so much smoother on my monitor 🙏🏼 I realize this can differ with frame rate and hardware, but I don’t see any lag or jumping anymore!
    - I can also double click an image to zoom in - bonus feature, and it also works with double tap on a touchscreen!

eatyourgreens commented:
I’ve got a workflow that uses the data-image viewer, though it was never live. You should be able to test that viewer here:

eatyourgreens commented:
https://github.com/zooniverse/front-end-monorepo/pull/4610 has examples for various data subject viewers, except `DataImageViewer`. I can’t remember why it was left out.

eatyourgreens commented:
>In the DataImageViewer, there’s a Story for Pan Zoom where on production the image toolbar buttons can be enabled for the image. However, on this branch the toolbar buttons do not zoom or reset the image in the viewer. The rotate button does work though.

Sorry, this is probably a very obvious question, but did you click Enable Zoom, above the image, before zooming in? The chart and the image have to be enabled individually, since there's only one toolbar to control zoom for both.
https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/story/subject-viewers-dataimageviewer--pan-zoom

goplayoutside3 commented:
> did you click Enable Zoom, above the image, before zooming in? The chart and the image have to be enabled individually

Yes, when zoom is enabled for the image, the rotate buttons works, but the zoom buttons do not affect the image at all. If I pan the image by click and drag, the reset button does not reset its position.

eatyourgreens commented:
This is really weird. I have the opposite problem in the storybook on this branch. Zoom works for the righthand image, even when I've clicked Disable Zoom to turn it off.


https://github.com/user-attachments/assets/564c1e2d-fe06-4951-91f4-51e759eb4c7d

I'm not 100% sure how to reproduce that. The chart zoom, on the left, uses `VisXZoom` and I don't know if having two instances of `VisXZoom` on the page might cause problems.

eatyourgreens commented:
Here's another screen recording. For some reason, enabling/disabling zoom for the lightcurve turns zoom on for the galaxy image. I hope this is helpful.


https://github.com/user-attachments/assets/8aab5920-060d-466a-9ed5-63c910cbd845

mcbouslog commented:
A brief update that I'll expand on more this afternoon:
- fixed issue with SeparateFrame toolbars impacting other frames
- fixed touchscreen dragging with touchEvents in VisXZoom
- fixed DataImageViewer issues, related to previous SeparateFrame refactors, both I think are working as expected now

Looking into now:
- can't touchMove drawn mark when creating or editing as expected

mcbouslog commented:
My mistake regarding touch events, I had not fixed touch dragging, and there wasn't an issue with drawn marks 🤦‍♂️ . I was using the Chrome toggle device toolbar dev tool and it was incorrect. Using my personal device and BrowserStack I've reverted or updated related changes and I think image dragging with touch and drawing with touch are working as expected across devices.

mcbouslog commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.js:
```diff
@@ -1,112 +1,132 @@
+          <title id={title.id}>{title.text}</title>
+        )}
+        <svg
+          style={{ touchAction: move ? 'none' : 'pinch-zoom' }}
```
I'm not 100% sure about `touch-action: none` if in `move` mode, but the UX I'm addressing with that is when in `move` mode, if I start to pinch to zoom, it often starts dragging briefly before the pinch registers and the image jumps around. With `touch-action: none` in `move` mode a user can't pinch to zoom, which might be unexpected, but the toolbar buttons will zoom, and going to `annotate` mode allows pinch-zoom, so that's the reasoning.

goplayoutside3 replied:
Move mode on a touchscreen works much better! It's a smooth drag experience I can use the image toolbar to zoom and rotate.

I actually don't think that annotate mode need pinch-zoom, do you? For instance, when in annotate mode and I try to pinch the viewer, it draws a mark on my phone anyway. Can we use `touchAction: 'none'` for both modes?

mcbouslog replied:
Yes I agree, I use the toolbar zoom buttons in annotate mode, and felt like the pinch zoom in annotate mode wasn't ideal (creates a mark like you note, and once zoomed in if I want to pan to make another mark, I have to pinch zoom out, then toggle to move mode, then pan, compared to using toolbar zoom buttons the toolbar and notably pan/zoom button stays in view).

I've updated to just `touchAction: 'none'` and we can revisit if there's feedback otherwise.

mcbouslog commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SVGComponents/VisXZoom/VisXZoom.js:
```diff
@@ -91,27 +95,32 @@ function VisXZoom({
+
+    if (!zoomRef.current.isDragging && !panning) return
+    
+    zoomRef.current.dragEnd()
```
This is slightly different behavior than the current image viewer. With the current image viewer, if a user drags an image out of the viewer it will continue to drag. With this change when the pointer leaves the viewer the image stops dragging. As a user, I could see how this _might_ be inconvenient, but it prevents a bug/poor UI in the current viewer where if a user drags then releases then the pointer re-enters the viewer it is still dragging and the image jumps around.

It's certainly debatable, but I think this new UI is preferable.

goplayoutside3 replied:
Agreed this UI is preferable. I think users of FEM projects are used to having to drag in excess to get the current images to move because of the lag bug (https://github.com/zooniverse/front-end-monorepo/issues/5110), but they'll be able to adjust to this new UX.

Related: I think something to remember is that you _can_ drag a drawn mark outside of the viewer. I think that happens because a mark can be placed anywhere on the image (or video) regardless if viewable at current zoom/rotate/pan levels. Post-merge of this PR I'm interested in discussing the implications of, for example, accidentally dragging a point outside of the viewer and then submitting the classification unaware of the stray point. Small touchscreens are at the highest risk because a drawing workflow defaults to 'annotate' mode and it's easy to try to scroll down while touching the subject viewer.

eatyourgreens replied:
Oops, that should have been:

https://github.com/zooniverse/front-end-monorepo/blob/a43d853c0274fd776bf2b8a220557a2da747bd5c/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.js#L62-L68

goplayoutside3 replied:
When I view this drawing + image [workflow](https://frontend.preview.zooniverse.org/projects/markb-panoptes/test-project-1-mb-fem-lab/classify/workflow/3847?env=staging) on my phone, I can easily draw a bunch of marks and then drag them outside of the viewer. The task area indicates I've drawn a non-zero number of marks even though I can't see them.

Also true of this drawing + video [workflow](https://frontend.preview.zooniverse.org/projects/markb-panoptes/test-project-1-mb-fem-lab/classify/workflow/3848?env=staging). Can't zoom out on a video either.

eatyourgreens replied:
You should open an issue for that. 

Here I create and delete two marks on my phone, by dragging them off the canvas. 


https://github.com/user-attachments/assets/bcf449e7-e628-4af8-b0ec-0ed48ce4547d

eatyourgreens replied:
Here’s a video of the video subject workflow, where I draw a circle then delete it by dragging it off the canvas. 


https://github.com/user-attachments/assets/655b0c1f-f1de-45e2-8f73-9857bdffc6af

eatyourgreens replied:
> Can't zoom out on a video either.

Well, no, no browser supports zooming into or out of a video. 

You could perhaps embed the video tag into the embedded SVG, as an external object.

mcbouslog replied:
Thank you @eatyourgreens for opening the related #6745 .

mcbouslog commented:
Note: to test the DataImageViewer in app-project, I've been using https://local.zooniverse.org:3000/projects/mrniaboc/superluminous-supernovae/classify/workflow/16264?env=production&demo=true .

goplayoutside3 reviewed (approved):
Approved! I think this is ready to merge, and to be safe we can request additional testing of frontend.preview on Mon/Tues next week.

One thing I noticed which is non-blocking to merging:
- Removal of the placeholder image behind the subject removes any indication of the left and top borders of the viewer. 
- The placeholder was intentionally displayed behind the subject to address https://github.com/zooniverse/front-end-monorepo/issues/1281 when a subject is rotated or zoomed-out enough to show the empty space. 
- For this PR, I recommend adding a subtle background color behind a subject with contrast to the page background (in both light and dark mode). We can double check with Sean when he's back post-merge.

mcbouslog commented:
@goplayoutside3 great note about the background color. I've added `light-4` which is close to the placeholder color for now, and indeed let's revisit with Sean.
